src/ipvs/ip_vs_redirect.c

```diff
@@ -0,0 +1,409 @@
+        if (dp_vs_redirect_ring[cid][peer_cid]) {
+            nb_rb = rte_ring_dequeue_burst(dp_vs_redirect_ring[cid][peer_cid],
+                                           (void**)mbufs,
+                                           NETIF_MAX_PKT_BURST, NULL);
+            if (nb_rb > 0) {
+                lcore_process_packets(qconf, mbufs, cid, nb_rb, 1);
+            }
+        }
+    }
+}
+
+static int dp_vs_redirect_table_create(void)
+{
+    int i;
+    char poolname[32];
+
+    /*
+     * allocate redirect cache on each NUMA socket and its size is
+     * same as conn_pool_size
+     */
+    for (i = 0; i < get_numa_nodes(); i++) {
+        snprintf(poolname, sizeof(poolname), "dp_vs_redirect_%d", i);
+        dp_vs_cr_cache[i] =
+            rte_mempool_create(poolname,
+                               dp_vs_conn_pool_size(),
+                               sizeof(struct dp_vs_redirect),
+                               dp_vs_conn_pool_cache_size(),
+                               0, NULL, NULL, NULL, NULL,
+                               i, 0);
+        if (!dp_vs_cr_cache[i]) {
+            return EDPVS_NOMEM;
+        }
+    }
+
+    /* allocate the global redirect hash table, per socket? */
+    dp_vs_cr_tbl =
+        rte_malloc_socket(NULL, sizeof(struct list_head ) * DPVS_CR_TBL_SIZE,
+                          RTE_CACHE_LINE_SIZE, rte_socket_id());
+    if (!dp_vs_cr_tbl) {
+        return EDPVS_NOMEM;
+    }
+
+    /* init the global redirect hash table */
+    for (i = 0; i < DPVS_CR_TBL_SIZE; i++) {
+        INIT_LIST_HEAD(&dp_vs_cr_tbl[i]);
+        rte_spinlock_init(&dp_vs_cr_lock[i]);
+    }
+
+    return EDPVS_OK;
+}
+
+static void dp_vs_redirect_table_free(void)
+{
+    int i;
+
+    for (i = 0; i < get_numa_nodes(); i++) {
+        rte_mempool_free(dp_vs_cr_cache[i]);
+    }
+
+    /* release the global redirect hash table */
+    if (dp_vs_cr_tbl) {
+        rte_free(dp_vs_cr_tbl);
+    }
+}
+
+/*
+ * Each lcore allocates redirect rings with the other lcores espectively.
+ */
+static int dp_vs_redirect_ring_create(void)
+{
+    char name_buf[RTE_RING_NAMESIZE];
+    int socket_id;
+    lcoreid_t cid, peer_cid;
+
+    socket_id = rte_socket_id();
+
+    for (cid = 0; cid < DPVS_MAX_LCORE; cid++) {
+        if (cid == rte_get_master_lcore() || !rte_lcore_is_enabled(cid)) {
+            continue;
+        }
+
+        for (peer_cid = 0; peer_cid < DPVS_MAX_LCORE; peer_cid++) {
+            if (!rte_lcore_is_enabled(peer_cid)
+                || peer_cid == rte_get_master_lcore()
+                || cid == peer_cid) {
+                continue;
+            }
+
+            snprintf(name_buf, RTE_RING_NAMESIZE,
+                     "dp_vs_redirect_ring[%d[%d]", cid, peer_cid);
+
+            dp_vs_redirect_ring[cid][peer_cid] =
+                rte_ring_create(name_buf, DPVS_REDIRECT_RING_SIZE, socket_id,
+                                RING_F_SP_ENQ | RING_F_SC_DEQ);
```

Comment:
Can we use "multi-producer  single-consumer" RTE Ring ? If so, the two-dimentional ring array can be reduced to a linear array.

Reply:
The reason is for the performance consideration to avoid multiple cores contending the same ring to enqueue the packets. And the side effect is the creation of Nx(N-1) rings which is bad for scalability.